drivers/regulator/regulator_nxp_vref.c

```diff
@@ -0,0 +1,272 @@
+	}
+
+	/* Enable output buffer */
+	*csr |= VREF_CSR_BUF21EN_MASK;
+
+	return 0;
+}
+
+static int regulator_nxp_vref_disable(const struct device *dev)
+{
+	const struct regulator_nxp_vref_config *config = dev->config;
+	const VREF_Type *base = (VREF_Type *) config->base;
+
+	volatile uint32_t *const csr = (volatile uint32_t *const) &base->CSR;
+
+	/*
+	 * Disable HC Bandgap, LP Bandgap, and Buf21
+	 * to achieve "Off" mode of VREF
+	 */
+	*csr &= ~VREF_CSR_HCBGEN_MASK;
+	*csr &= ~VREF_CSR_LPBGEN_MASK;
+
+	return 0;
+}
+
+static int regulator_nxp_vref_set_mode(const struct device *dev, regulator_mode_t mode)
+{
+	const struct regulator_nxp_vref_config *config = dev->config;
+	const VREF_Type *base = (VREF_Type *) config->base;
+
+	volatile uint32_t *const csr = (volatile uint32_t *const) &base->CSR;
+
+	if (mode == NXP_VREF_MODE_STANDBY) {
+		*csr &= ~VREF_CSR_REGEN_MASK &
+			~VREF_CSR_CHOPEN_MASK &
+			~VREF_CSR_HI_PWR_LV_MASK &
+			~VREF_CSR_BUF21EN_MASK;
+			return 0;
+	}
+
+	if (mode == NXP_VREF_MODE_LOW_POWER) {
+		*csr &= ~VREF_CSR_REGEN_MASK &
+			~VREF_CSR_CHOPEN_MASK &
+			~VREF_CSR_HI_PWR_LV_MASK;
+		*csr |= VREF_CSR_BUF21EN_MASK;
+	} else if (mode == NXP_VREF_MODE_HIGH_POWER) {
+		*csr &= ~VREF_CSR_REGEN_MASK &
+			~VREF_CSR_CHOPEN_MASK;
+		*csr |= VREF_CSR_HI_PWR_LV_MASK &
+			VREF_CSR_BUF21EN_MASK;
+	} else if (mode == NXP_VREF_MODE_INTERNAL_REGULATOR) {
+		*csr |= VREF_CSR_REGEN_MASK &
+			 VREF_CSR_CHOPEN_MASK &
+			 VREF_CSR_HI_PWR_LV_MASK &
+			 VREF_CSR_BUF21EN_MASK;
+	} else {
+		return -EINVAL;
+	}
+
+	k_busy_wait(config->buf_start_delay);
+
+	return 0;
+}
+
+static int regulator_nxp_vref_get_mode(const struct device *dev, regulator_mode_t *mode)
+{
+	const struct regulator_nxp_vref_config *config = dev->config;
+	const VREF_Type *base = (VREF_Type *) config->base;
+
+	volatile uint32_t *const csr = (volatile uint32_t *const) &base->CSR;
+
+	if (*csr & VREF_CSR_REGEN_MASK) {
+		*mode = NXP_VREF_MODE_INTERNAL_REGULATOR;
+	} else if (*csr & VREF_CSR_HI_PWR_LV_MASK) {
+		*mode = NXP_VREF_MODE_HIGH_POWER;
+	} else if (*csr & VREF_CSR_BUF21EN_MASK) {
+		*mode = NXP_VREF_MODE_LOW_POWER;
+	} else {
+		*mode = NXP_VREF_MODE_STANDBY;
+	}
+
+	return 0;
+}
+
+static inline unsigned int regulator_nxp_vref_count_voltages(const struct device *dev)
+{
+	return ARRAY_SIZE(utrim_vals);
+}
+
+static int regulator_nxp_vref_list_voltage(const struct device *dev,
+						unsigned int idx, int32_t *volt_uv)
+{
+	if (idx > ARRAY_SIZE(utrim_vals)) {
+		return -EINVAL;
+	}
+
+	*volt_uv = utrim_vals[idx].voltage;
+
+	return 0;
+}
+
+static int regulator_nxp_vref_set_voltage(const struct device *dev,
+					int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_nxp_vref_config *config = dev->config;
+	const VREF_Type *base = (VREF_Type *) config->base;
+
+	volatile uint32_t *const utrim = (volatile uint32_t *const) &base->UTRIM;
+	int32_t volt_uv;
+
+	if (!(min_uv > NXP_VREF_MIN_UV && max_uv < NXP_VREF_MAX_UV && max_uv > min_uv)) {
```

Comment:
drivers should not check for `max_uv > min_uv`.

Reply:
why not? I dont see where else it is done

Reply:
Use linear range to check if requested voltages are valid.
Idx will return the required register value to set.

```suggestion
	int ret;
	uint8_t idx;

	ret = linear_range_get_win_index(&vref_range, min_uv, max_uv, &idx);

	if (ret == -EINVAL) {
		return ret;
	}
```